File: tests/test_storage_migrator.py
"""Tests for dispatch/storage_migrator.py — JSON/JSONL → SQLite migration."""

import json
import sys
import tempfile
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

import unittest
from harness_core.dispatch.durable_store import DurableStore
from harness_core.dispatch.storage_migrator import (
    FullMigrationReport,
    MigrationReport,
    STORAGE_MIGRATOR_SCHEMA_VERSION,
    _read_json_file,
    _read_jsonl_file,
    full_migration,
    migrate_events_jsonl_to_sqlite,
    migrate_plans_json_to_sqlite,
    migrate_repos_json_to_sqlite,
)


class MigrationReportTests(unittest.TestCase):
    def test_fields(self):
        r = MigrationReport(
            source="a.json", target="sqlite",
            records_migrated=5, errors=["e1"], duration_ms=10.0,
        )
        self.assertEqual(r.records_migrated, 5)
        self.assertEqual(r.source, "a.json")

    def test_immutable(self):
        r = MigrationReport(source="a", target="b", records_migrated=0, errors=[], duration_ms=0.0)
        with self.assertRaises(AttributeError):
            r.records_migrated = 1  # type: ignore[misc]


class FullMigrationReportTests(unittest.TestCase):
    def test_fields(self):
        r = FullMigrationReport(
            plans=MigrationReport("a", "b", 1, [], 1.0),
            repos=MigrationReport("c", "d", 2, [], 2.0),
            events=MigrationReport("e", "f", 3, [], 3.0),
            total_duration_ms=6.0,
        )
        self.assertEqual(r.plans.records_migrated, 1)
        self.assertEqual(r.total_duration_ms, 6.0)


class ReadJsonFileTests(unittest.TestCase):
    def test_read_valid(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as f:
            json.dump({"key": "value"}, f)
            f.flush()
            result = _read_json_file(Path(f.name))
        self.assertEqual(result, {"key": "value"})
        Path(f.name).unlink(missing_ok=True)

    def test_read_nonexistent(self):
        self.assertIsNone(_read_json_file(Path("/nonexistent/file.json")))

    def test_read_invalid_json(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as f:
            f.write("not json{{{")
            f.flush()
            result = _read_json_file(Path(f.name))
        self.assertIsNone(result)
        Path(f.name).unlink(missing_ok=True)


class ReadJsonlFileTests(unittest.TestCase):
    def test_read_valid(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".jsonl", delete=False) as f:
            f.write('{"id": "1"}\n{"id": "2"}\n')
            f.flush()
            records, errors = _read_jsonl_file(Path(f.name))
        self.assertEqual(len(records), 2)
        self.assertEqual(errors, [])
        Path(f.name).unlink(missing_ok=True)

    def test_read_nonexistent(self):
        records, errors = _read_jsonl_file(Path("/nonexistent/file.jsonl"))
        self.assertEqual(records, [])
        self.assertEqual(errors, [])

    def test_read_with_blank_lines(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".jsonl", delete=False) as f:
            f.write('{"id": "1"}\n\n{"id": "2"}\n')
            f.flush()
            records, errors = _read_jsonl_file(Path(f.name))
        self.assertEqual(len(records), 2)
        self.assertEqual(errors, [])
        Path(f.name).unlink(missing_ok=True)

    def test_read_with_invalid_line(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".jsonl", delete=False) as f:
            f.write('{"id": "1"}\nnot json\n{"id": "2"}\n')
            f.flush()
            records, errors = _read_jsonl_file(Path(f.name))
        self.assertEqual(len(records), 2)
        self.assertEqual(len(errors), 1)
        self.assertIn("line 2", errors[0])
        Path(f.name).unlink(missing_ok=True)

    def test_read_multiple_invalid_lines(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".jsonl", delete=False) as f:
            f.write('not json 1\n{"id": "1"}\nnot json 2\n')
            f.flush()
            records, errors = _read_jsonl_file(Path(f.name))
        self.assertEqual(len(records), 1)
        self.assertEqual(len(errors), 2)
        self.assertIn("line 1", errors[0])
        self.assertIn("line 3", errors[1])
        Path(f.name).unlink(missing_ok=True)


class MigratePlansTests(unittest.TestCase):
    def setUp(self):
        self.store = DurableStore()

    def tearDown(self):
        self.store.close()

    def test_migrate_valid_plans(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as f:
            json.dump({
                "schema_version": "app_plans.v1",
                "plans": [
                    {"plan_id": "p1", "task": {"objective": "build"}, "schema_version": "resource_plan.v1"},
                    {"plan_id": "p2", "task": {"objective": "test"}},
                ],
            }, f)
            f.flush()
            report = migrate_plans_json_to_sqlite(Path(f.name), self.store)
        self.assertEqual(report.records_migrated, 2)
        self.assertEqual(report.errors, [])
        self.assertIsNotNone(self.store.get_plan("p1"))
        self.assertIsNotNone(self.store.get_plan("p2"))
        Path(f.name).unlink(missing_ok=True)

    def test_migrate_missing_plan_id(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as f:
            json.dump({"plans": [{"task": "no id"}]}, f)
            f.flush()
            report = migrate_plans_json_to_sqlite(Path(f.name), self.store)
        self.assertEqual(report.records_migrated, 0)
        self.assertEqual(len(report.errors), 1)
        Path(f.name).unlink(missing_ok=True)

    def test_migrate_nonexistent_file(self):
        report = migrate_plans_json_to_sqlite(Path("/nonexistent.json"), self.store)
        self.assertEqual(report.records_migrated, 0)
        self.assertIn("file not found", report.errors[0])

    def test_migrate_preserves_schema_version(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as f:
            json.dump({"plans": [{"plan_id": "p1", "schema_version": "custom.v2"}]}, f)
            f.flush()
            migrate_plans_json_to_sqlite(Path(f.name), self.store)
        plan = self.store.get_plan("p1")
        self.assertEqual(plan.schema_version, "custom.v2")  # type: ignore[union-attr]
        Path(f.name).unlink(missing_ok=True)


class MigrateReposTests(unittest.TestCase):
    def setUp(self):
        self.store = DurableStore()

    def tearDown(self):
        self.store.close()

    def test_migrate_valid_repos(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as f:
            json.dump({
                "repos": [
                    {"id": "r1", "name": "repo-1", "kind": "local"},
                    {"id": "r2", "name": "repo-2", "kind": "remote"},
                ],
            }, f)
            f.flush()
            report = migrate_repos_json_to_sqlite(Path(f.name), self.store)
        self.assertEqual(report.records_migrated, 2)
        self.assertEqual(report.errors, [])
        self.assertIsNotNone(self.store.get_repo("r1"))
        Path(f.name).unlink(missing_ok=True)

    def test_migrate_missing_repo_id(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as f:
            json.dump({"repos": [{"name": "no id"}]}, f)
            f.flush()
            report = migrate_repos_json_to_sqlite(Path(f.name), self.store)
        self.assertEqual(report.records_migrated, 0)
        Path(f.name).unlink(missing_ok=True)

    def test_migrate_nonexistent_file(self):
        report = migrate_repos_json_to_sqlite(Path("/nonexistent.json"), self.store)
        self.assertEqual(report.records_migrated, 0)


class MigrateEventsTests(unittest.TestCase):
    def setUp(self):
        self.store = DurableStore()

    def tearDown(self):
        self.store.close()

    def test_migrate_valid_events(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".jsonl", delete=False) as f:
            f.write(json.dumps({"event_id": "e1", "event_type": "deploy", "schema_version": "event.v1"}) + "\n")
            f.write(json.dumps({"event_id": "e2", "event_type": "build"}) + "\n")
            f.flush()
            report = migrate_events_jsonl_to_sqlite(Path(f.name), self.store)
        self.assertEqual(report.records_migrated, 2)
        self.assertEqual(report.errors, [])
        self.assertIsNotNone(self.store.get_event("e1"))
        Path(f.name).unlink(missing_ok=True)

    def test_migrate_missing_event_id(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".jsonl", delete=False) as f:
            f.write(json.dumps({"event_type": "deploy"}) + "\n")
            f.flush()
            report = migrate_events_jsonl_to_sqlite(Path(f.name), self.store)
        self.assertEqual(report.records_migrated, 0)
        Path(f.name).unlink(missing_ok=True)

    def test_migrate_nonexistent_file(self):
        report = migrate_events_jsonl_to_sqlite(Path("/nonexistent.jsonl"), self.store)
        self.assertEqual(report.records_migrated, 0)

    def test_migrate_empty_file(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".jsonl", delete=False) as f:
            f.flush()
            report = migrate_events_jsonl_to_sqlite(Path(f.name), self.store)
        self.assertEqual(report.records_migrated, 0)
        self.assertEqual(report.errors, [])
        Path(f.name).unlink(missing_ok=True)

    def test_migrate_with_malformed_lines_reports_errors(self):
        with tempfile.NamedTemporaryFile(mode="w", suffix=".jsonl", delete=False) as f:
            f.write('{"event_id": "e1"}\n')
            f.write('not valid json {{{\n')
            f.write('{"event_id": "e2"}\n')
            f.flush()
            report = migrate_events_jsonl_to_sqlite(Path(f.name), self.store)
        self.assertEqual(report.records_migrated, 2)
        self.assertEqual(len(report.errors), 1)
        self.assertIn("line 2", report.errors[0])
        Path(f.name).unlink(missing_ok=True)


class FullMigrationTests(unittest.TestCase):
    def test_full_migration(self):
        store = DurableStore()
        with tempfile.TemporaryDirectory() as tmpdir:
            plans_path = Path(tmpdir) / "plans.json"
            repos_path = Path(tmpdir) / "repos.json"
            events_path = Path(tmpdir) / "events.jsonl"

            plans_path.write_text(json.dumps({
                "plans": [{"plan_id": "p1", "task": {"obj": "a"}}],
            }))
            repos_path.write_text(json.dumps({
                "repos": [{"id": "r1", "name": "repo"}],
            }))
            events_path.write_text(json.dumps({
                "event_id": "e1", "event_type": "created",
            }) + "\n")

            report = full_migration(plans_path, repos_path, events_path, store)

        self.assertIsInstance(report, FullMigrationReport)
        self.assertEqual(report.plans.records_migrated, 1)
        self.assertEqual(report.repos.records_migrated, 1)
        self.assertEqual(report.events.records_migrated, 1)
        self.assertGreater(report.total_duration_ms, 0)
        store.close()

    def test_full_migration_with_errors(self):
        store = DurableStore()
        with tempfile.TemporaryDirectory() as tmpdir:
            plans_path = Path(tmpdir) / "plans.json"
            repos_path = Path(tmpdir) / "repos.json"
            events_path = Path(tmpdir) / "events.jsonl"

            plans_path.write_text(json.dumps({
                "plans": [{"task": "no plan_id"}],
            }))
            repos_path.write_text(json.dumps({"repos": []}))
            events_path.write_text("")

            report = full_migration(plans_path, repos_path, events_path, store)

        self.assertEqual(report.plans.records_migrated, 0)
        self.assertTrue(len(report.plans.errors) > 0)
        store.close()

    def test_full_migration_nonexistent_files(self):
        store = DurableStore()
        report = full_migration(
            Path("/no/plans.json"),
            Path("/no/repos.json"),
            Path("/no/events.jsonl"),
            store,
        )
        self.assertEqual(report.plans.records_migrated, 0)
        self.assertEqual(report.repos.records_migrated, 0)
        self.assertEqual(report.events.records_migrated, 0)
        store.close()


class SchemaVersionTests(unittest.TestCase):
    def test_schema_version_defined(self):
        self.assertEqual(STORAGE_MIGRATOR_SCHEMA_VERSION, "storage_migrator.v1")


if __name__ == "__main__":
    unittest.main()

File: engine/tests/test_storage_migrator.rs
use engine::storage::durable_store::DurableStore;
use engine::storage::storage_migrator::*;
use std::fs;
use tempfile::TempDir;

#[test]
fn test_migrate_plans_json_to_sqlite() {
    let tmp = TempDir::new().unwrap();
    let json_path = tmp.path().join("plans.json");
    fs::write(
        &json_path,
        r#"{"plans": [{"plan_id": "p1", "name": "test", "schema_version": "plan.v1"}]}"#,
    )
    .unwrap();

    let store = DurableStore::new_memory().unwrap();
    let report = migrate_plans_json_to_sqlite(&json_path, &store, 0.0, 0.1);
    assert_eq!(report.records_migrated, 1);
    assert!(report.errors.is_empty());

    let plan = store.get_plan("p1").unwrap().unwrap();
    assert_eq!(plan.data["name"], "test");
}

#[test]
fn test_migrate_plans_missing_file() {
    let store = DurableStore::new_memory().unwrap();
    let report = migrate_plans_json_to_sqlite(
        std::path::Path::new("/nonexistent/plans.json"),
        &store,
        0.0,
        0.1,
    );
    assert_eq!(report.records_migrated, 0);
    assert!(!report.errors.is_empty());
}

#[test]
fn test_migrate_plans_missing_id() {
    let tmp = TempDir::new().unwrap();
    let json_path = tmp.path().join("plans.json");
    fs::write(&json_path, r#"{"plans": [{"name": "no-id"}]}"#).unwrap();

    let store = DurableStore::new_memory().unwrap();
    let report = migrate_plans_json_to_sqlite(&json_path, &store, 0.0, 0.1);
    assert_eq!(report.records_migrated, 0);
    assert!(report.errors.iter().any(|e| e.contains("missing plan_id")));
}

#[test]
fn test_migrate_repos_json_to_sqlite() {
    let tmp = TempDir::new().unwrap();
    let json_path = tmp.path().join("repos.json");
    fs::write(
        &json_path,
        r#"{"repos": [{"id": "r1", "url": "https://github.com/test", "schema_version": "repo.v1"}]}"#,
    )
    .unwrap();

    let store = DurableStore::new_memory().unwrap();
    let report = migrate_repos_json_to_sqlite(&json_path, &store, 0.0, 0.1);
    assert_eq!(report.records_migrated, 1);
    assert!(report.errors.is_empty());

    let repo = store.get_repo("r1").unwrap().unwrap();
    assert_eq!(repo.data["url"], "https://github.com/test");
}

#[test]
fn test_migrate_events_jsonl_to_sqlite() {
    let tmp = TempDir::new().unwrap();
    let jsonl_path = tmp.path().join("events.jsonl");
    fs::write(
        &jsonl_path,
        r#"{"event_id": "e1", "event_type": "dispatch", "schema_version": "event.v1"}
{"event_id": "e2", "event_type": "error", "schema_version": "event.v1"}"#,
    )
    .unwrap();

    let store = DurableStore::new_memory().unwrap();
    let report = migrate_events_jsonl_to_sqlite(&jsonl_path, &store, 0.0, 0.1);
    assert_eq!(report.records_migrated, 2);
    assert!(report.errors.is_empty());

    let events = store.get_events(None, 100).unwrap();
    assert_eq!(events.len(), 2);
}

#[test]
fn test_migrate_events_missing_file() {
    let store = DurableStore::new_memory().unwrap();
    let report = migrate_events_jsonl_to_sqlite(
        std::path::Path::new("/nonexistent/events.jsonl"),
        &store,
        0.0,
        0.1,
    );
    assert_eq!(report.records_migrated, 0);
}

#[test]
fn test_migrate_events_bad_json() {
    let tmp = TempDir::new().unwrap();
    let jsonl_path = tmp.path().join("events.jsonl");
    fs::write(
        &jsonl_path,
        r#"{"event_id": "e1", "event_type": "good"}
not valid json
{"event_id": "e2", "event_type": "also-good"}"#,
    )
    .unwrap();

    let store = DurableStore::new_memory().unwrap();
    let report = migrate_events_jsonl_to_sqlite(&jsonl_path, &store, 0.0, 0.1);
    assert_eq!(report.records_migrated, 2);
    assert_eq!(report.errors.len(), 1);
    assert!(!report.errors.is_empty());
}

#[test]
fn test_full_migration() {
    let tmp = TempDir::new().unwrap();
    fs::write(
        tmp.path().join("plans.json"),
        r#"{"plans": [{"plan_id": "p1", "schema_version": "plan.v1"}]}"#,
    )
    .unwrap();
    fs::write(
        tmp.path().join("repos.json"),
        r#"{"repos": [{"id": "r1", "schema_version": "repo.v1"}]}"#,
    )
    .unwrap();
    fs::write(
        tmp.path().join("events.jsonl"),
        r#"{"event_id": "e1", "schema_version": "event.v1"}"#,
    )
    .unwrap();

    let store = DurableStore::new_memory().unwrap();
    let report = full_migration(
        &tmp.path().join("plans.json"),
        &tmp.path().join("repos.json"),
        &tmp.path().join("events.jsonl"),
        &store,
        0.0,
        0.1,
    );
    assert_eq!(report.plans.records_migrated, 1);
    assert_eq!(report.repos.records_migrated, 1);
    assert_eq!(report.events.records_migrated, 1);

    let stats = store.stats().unwrap();
    assert_eq!(stats["plans"], 1);
    assert_eq!(stats["repos"], 1);
    assert_eq!(stats["events"], 1);
}

#[test]
fn test_migration_report_serde() {
    let report = MigrationReport {
        source: "test.json".to_string(),
        target: "sqlite".to_string(),
        records_migrated: 5,
        errors: vec![],
        duration_ms: 100.0,
    };
    let json = serde_json::to_string(&report).unwrap();
    let r: MigrationReport = serde_json::from_str(&json).unwrap();
    assert_eq!(r.records_migrated, 5);
}
